### PR TITLE
Buildpack is converted to a supply buildpack

### PR DESCRIPTION
**Pull request by izgeri**
Connected to #17 

In order to convert our buildpacker to a supply buildpack, this PR moves the buildpack installation logic into a single `supply` script and changes the directories in which dependencies are installed, along with a few other changes to support this shift. It also cleans up tests a bit and adds comments where useful, and updates the `conjur-env` binary to use updated dependencies and Go modules.

### CODE REVIEW

**Comment by izgeri** [comment]
I added some notes to the [original issue](https://github.com/cyberark/cloudfoundry-conjur-buildpack/issues/17#issuecomment-461265751) on the work remaining on this PR

File: bin/compile [hunk 1]
```diff
@@ -1,14 +1,5 @@
...
-BUILD_DIR=$1
-BIN_DIR=$(cd $(dirname $0); pwd)
-BUILDPACK_DIR=$(dirname $BIN_DIR)
-
-pushd ${BUILD_DIR}
-  mkdir -p .profile.d vendor
-  cp ${BUILDPACK_DIR}/vendor/conjur-env ./vendor/conjur-env
-  cp ${BUILDPACK_DIR}/lib/0001_retrieve-secrets.sh ./.profile.d/0001_retrieve-secrets.sh
-popd
+exit
```
Comment (micahlee) [inline bin/compile]: Would you mind adding the newline here at the end?

File: bin/supply [hunk 2]
```diff
@@ -0,0 +1,36 @@
+#!/bin/bash -e
+# bin/supply <build-dir> <cache-dir> <deps-dir> <index>
+
+BUILD_DIR=$1
```
Comment (micahlee) [inline bin/supply]: It would be helpful to add a quick comment for each of the directories and their purpose/role, either here or in the README.

**Comment by izgeri** [comment]
Added #20 for adding automated tests to run against the live PCF instance

File: ci/features/buildpackless.feature [hunk 3]
```diff
@@ -3,7 +3,9 @@ Feature: profile d scripts without the buildpack
 
   @BUILD_DIR
   Scenario: Populates environment with secrets from Conjur
-    Given the compile script is run against the app's root folder
+    Given the build directory has a secrets.yml file
+    And VCAP_SERVICES contains cyberark-conjur credentials
+    And the supply script is run against the app's root folder
```
Comment (izgeri) [inline ci/features/buildpackless.feature]: Should we also have two tests that validate what happens when
- the build dir does not have secrets.yml, and/or
- VCAP_SERVICES does not contain cyberark-conjur creds?

File: ci/features/compile.feature [hunk 4]
```diff
@@ -1,9 +0,0 @@
-Feature: Compile script
```
Comment (izgeri) [inline ci/features/compile.feature]: Should we leave a test here that validates that this script exits for now? I could go either way.

File: ci/features/profile_d.feature [hunk 5]
```diff
@@ -3,7 +3,9 @@ Feature: profile d script
 
   @BUILD_DIR
   Scenario: Populates environment with secrets from Conjur
-    Given the 'compile' script is run
+    Given the build directory has a secrets.yml file
+    And VCAP_SERVICES contains cyberark-conjur credentials
+    And the supply script is run against the app's root folder
```
Comment (izgeri) [inline ci/features/profile_d.feature]: Same comment here, should we test when the conditions are not right?

File: ci/features/step_definitions/profile_d_steps.rb [hunk 6]
```diff
@@ -1,17 +1,10 @@
 When(/^the retrieve secrets \.profile\.d script is sourced$/) do
   @commands ||= []
   @commands << <<EOL
-. #{@BUILD_DIR}/.profile.d/0001_retrieve-secrets.sh #{@BUILD_DIR}
+HOME=#{@BUILD_DIR} DEPS_DIR=#{@DEPS_DIR} . #{@DEPS_DIR}/#{@INDEX_DIR}/.profile.d/0001_retrieve-secrets.sh
```
Comment (izgeri) [inline ci/features/step_definitions/profile_d_steps.rb]: Can we add a comment here explaining the reason for passing in HOME / DEPS_DIR and the syntax being used?

File: lib/0001_retrieve-secrets.sh [hunk 7]
```diff
@@ -25,16 +25,17 @@ trap 'err_report $LINENO' ERR
...
 declare xtrace=""
 case $- in
   (*x*) xtrace="xtrace";;
 esac
 set +x
 
-# inject secrets into environment
-pushd $1
-  eval "$(./vendor/conjur-env)"
+# $HOME points to the app directory, which should contains a secrets.yml file.
```
Comment (izgeri) [inline lib/0001_retrieve-secrets.sh]: Is there documentation we can link to that will share how we know HOME and DEPS_DIR will be set when this script is run?

File: bin/supply (hunk 2 above)
Comment (jtuttle) [reply bin/supply]: Done!

File: ci/features/step_definitions/profile_d_steps.rb (hunk 6 above)
Comment (jtuttle) [reply ci/features/step_definitions/profile_d_steps.rb]: Yep, added a comment.

File: lib/0001_retrieve-secrets.sh (hunk 7 above)
Comment (jtuttle) [reply lib/0001_retrieve-secrets.sh]: Unfortunately no. This is knowledge I gained from the examples linked to in Slack conversations. This seems to be quite poorly documented.

**Comment by izgeri** [comment]
@jtuttle can you revise the PR description so that it has more info about what the changes contain? might be helpful if looking back at this PR in the future

File: ci/features/buildpackless.feature (hunk 3 above)
Comment (jtuttle) [reply ci/features/buildpackless.feature]: These conditions are already tested in the `supply.feature` so I think we're okay.

File: ci/features/compile.feature (hunk 4 above)
Comment (jtuttle) [reply ci/features/compile.feature]: I'd say let's just leave it out. It's such a simple script that if anything goes wrong it'd probably be something not covered by the test we would right.

File: ci/features/profile_d.feature (hunk 5 above)
Comment (jtuttle) [reply ci/features/profile_d.feature]: Hm this feature appears to be identical to the buildpackless one. They use different step definitions on `master`, but they essentially boil down to the same thing. I think we should probably just remove the `buildpackless` one.

**Comment by jtuttle** [comment]
@izgeri PR description updated.

File: README.md [hunk 8]
```diff
@@ -1,4 +1,4 @@
-The CyberArk Conjur Buildpack is a [decorator buildpack](https://github.com/cf-platform-eng/meta-buildpack#what-is-a-decorator) that provides convenient and secure access to secrets stored in Conjur.
+The CyberArk Conjur Buildpack is a [supply buildpack](https://docs.cloudfoundry.org/buildpacks/custom.html#contract) that provides convenient and secure access to secrets stored in Conjur.
```
Comment (izgeri) [inline README.md]: I wonder if this is a better link?
https://docs.cloudfoundry.org/buildpacks/understand-buildpacks.html#supply-script

File: README.md [hunk 9]
```diff
@@ -1,4 +1,4 @@
...
 
 The buildpack carries out the following:
```
Comment (izgeri) [inline README.md]: I think more accurately the buildpack installs some scripts that do the following

File: README.md (hunk 8 above)
Comment (jtuttle) [reply README.md]: Indubitably.

File: README.md [hunk 10]
```diff
@@ -10,30 +10,20 @@ Internally, the Conjur Buildpack uses [Summon](https://cyberark.github.io/summon
...
-of the buildpack index and your application must automatically detect the appropriate buildpack on `cf push`. If your application requires you to specify a buildpack on `cf push`, you can follow the [instructions for apps with custom buildpacks](#custom-buildpack-usage) to benefit from the convenience of the Conjur Buildpack without having to rely on the `meta-buildpack` for it to be invoked.
-
-### Lifecycle scripts
-
-The buildpack is comprised of the [3 lifecycle scripts](https://github.com/cf-platform-eng/meta-buildpack#how-to-write-a-decorator) that are required for decorator buildpacks.
+The buildpack uses a [supply script](https://docs.cloudfoundry.org/buildpacks/custom.html#contract) to copy files into the application's dependency directory under a subdirectory corresponding to the buildpack's index. The `lib/0001_retrieve-secrets.sh` script is copied into a `.profile.d` subdirectory so that it will run automatically when the app starts and the `conjur-env` binary is copied to a `vendor` subdirectory. In other words, your application will end up with the following two files:
```
Comment (izgeri) [inline README.md]: https://docs.cloudfoundry.org/buildpacks/understand-buildpacks.html#supply-script is definitely a better link here

File: README.md [hunk 11]
```diff
@@ -10,30 +10,20 @@ Internally, the Conjur Buildpack uses [Summon](https://cyberark.github.io/summon
...
-+ Detect script always returns a non-zero exit status.
-+ Compile script copies the `conjur-env` binary into your application's `vendor` directory and the `lib/0001_retrieve-secrets.sh` script into your application's `.profile.d` directory.
-+ Decorate script returns a 0 when `secrets.yml` exists in the root folder of your app and the "cyberark-conjur" key is present in `VCAP_SERVICES`, otherwise non-zero exit status
+```
+- $DEPS_DIR/$BUILDPACK_INDEX/.profile.d/0001-retrieve-secrets.sh
+- $DEPS_DIR/$BUILDPACK_INDEX/vendor/conjur-env
+```
 
-The `.profile.d` script is responsible for retrieving secrets and injecting them into the session environment variables at the start of the app.
+The `.profile.d` script is run automatically when the application starts and is responsible for retrieving secrets and injecting them into the app's session environment variables.
```
Comment (izgeri) [inline README.md]: Maybe worth adding something here about how the `conjur-env` binary leverages [Conjur's Go API](https://github.com/cyberark/conjur-api-go) to open a connection to Conjur and retrieve secrets

File: README.md [hunk 12]
```diff
@@ -86,25 +68,6 @@ AWS_REGION: us-east-1
...
-Once you run `cf push`, assuming the app is bound to a Conjur service instance, the `meta-buildpack` will first invoke the appropriate language buildpack and then any decorator buildpacks (including the Conjur Buildpack). The secrets specified in the `secrets.yml` file will be available in the session environment variables at the start of the app.
-
-#### <a name="custom-buildpack-usage"></a>Usage for users with custom buildpacks
-
-To retrieve secrets without using `meta-buildpack` you can simply:
-
-+ copy the contents of the `./lib` directory in a release to the `./.profile.d` directory relative to your app's root directory.
-+ copy the contents of the `./vendor` directory in a release to the `./vendor` directory relative to your app's root directory.
-
 When your application starts the secrets specified in the `secrets.yml` file will now be available in the session environment variables at the start of the app.
```
Comment (izgeri) [inline README.md]: We may still have users who need to use the decorator buildpack functionality, and they can do so using the 1.0.0 release of the buildpack. I feel pretty strongly that we should continue to include those instructions somewhere - can we either do that here, or log an issue in the [conjur-docs](https://github.com/cyberark/conjur-docs) to add documentation for using the original meta-buildpack functionality to the (non-existent yet) CF buildpack documentation?

Comment (izgeri) [reply README.md]: Issue #21 will address this

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+
+- [Prerequisites](#prerequisites)
+- [Pull Request Workflow](#pull-request-workflow)
+- [Updating the `conjur-env` Binary](#updating-the-conjur-env-binary)
+- [Testing](#testing)
+- [Releasing](#releasing)
+
+## Prerequisites
+
+Before getting started, you should install some developer tools. These are not required to deploy the Conjur Service Broker but they will let you develop using a standardized, expertly configured environment.
```
Comment (micahlee) [inline CONTRIBUTING.md]: 😎

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+
+To run the test suite on your local machine:
+```
+$ ./package.sh   # Create the conjur-env binary in the vendor dir and a ZIP of the project contents
+$ ./test.sh      # Run the test suite
+```
+
+### Releasing
+
+1. Based on the unreleased content, determine the new version number and update the [VERSION](VERSION) file.
```
Comment (micahlee) [inline CONTRIBUTING.md]: Should there be a mention here about semver, or what it means to `determine the new version number`?

**Pull request merged**